### ISSUE TITLE
Transcoding

### ISSUE THREAD

**Issue by DDoubleU** [vishen/go-chromecast#17]
Hi,

I've used the Linux binaries and built the application on Windows as well and both work perfectly! Now I'm trying to play some files in MP4 and MKV container, but some of them are using codecs that are not supported by the Chromecast. I've installed FFMPEG for the transcoding, but what exactly "triggers" the transcoding? Container, codec? There was one MKV file which plays video, but no audio nor subtitles which makes me think it doesn't "call" FFMPEG. Is there any way to force transcoding?

Thanks in advance!

**Comment by vishen** [comment]
Hmm, I just tried on my Mac and Linux machines, and it seems to transcode `avi` and `mkv` files correctly to `mp4`. By default it should attempt to transcode the media to mp4 if it is an unknown media format (https://github.com/vishen/go-chromecast/blob/master/application/application.go#L569:576). 

If you run `go-chromecast --debug` with the debug flag does it give you anything about what the problem might be?

It runs the following command https://github.com/vishen/go-chromecast/blob/master/application/application.go#L698:706 with ffmpeg, maybe the command doesn't work with your ffmpeg?

What version of ffmpeg are you using? Mine is
```
ffmpeg -version                                                                                                                    15:10:40
ffmpeg version 3.4 Copyright (c) 2000-2017 the FFmpeg developers
built with Apple LLVM version 9.0.0 (clang-900.0.37)
configuration: --prefix=/usr/local/Cellar/ffmpeg/3.4 --enable-shared --enable-pthreads --enable-version3 --enable-hardcoded-tables --enable-avresample --cc=clang --host-cflags= --host-ldflags= --enable-gpl --enable-libmp3lame --enable-libvpx --enable-libx264 --enable-libxvid --enable-opencl --enable-videotoolbox --disable-lzma
libavutil      55. 78.100 / 55. 78.100
libavcodec     57.107.100 / 57.107.100
libavformat    57. 83.100 / 57. 83.100
libavdevice    57. 10.100 / 57. 10.100
libavfilter     6.107.100 /  6.107.100
libavresample   3.  7.  0 /  3.  7.  0
libswscale      4.  8.100 /  4.  8.100
libswresample   2.  9.100 /  2.  9.100
libpostproc    54.  7.100 / 54.  7.100
```

**Comment by DDoubleU** [comment]
If I run the command 'go-chromecast load "F:\Media\test.mkv" -n "Woonkamer" --debug' I get the following output:
`[36mINFO[0m[0000] [connection] (1)sender-0 -> receiver-0 [urn:x-cast:com.google.cast.tp.connection]: {"type":"CONNECT","requestId":1}
[36mINFO[0m[0000] [connection] (2)sender-0 -> receiver-0 [urn:x-cast:com.google.cast.receiver]: {"type":"GET_STATUS","requestId":2}
[36mINFO[0m[0000] [connection] (2)sender-0 <- receiver-0 [urn:x-cast:com.google.cast.receiver]: {"requestId":2,"status":{"applications":[{"appId":"E8C28D3C","displayName":"Backdrop","isIdleScreen":true,"launchedFromCloud":false,"namespaces":[{"name":"urn:x-cast:com.google.cast.debugoverlay"},{"name":"urn:x-cast:com.google.cast.cac"},{"name":"urn:x-cast:com.google.cast.sse"},{"name":"urn:x-cast:com.google.cast.remotecontrol"}],"sessionId":"4494b3d4-37ba-4d2c-9cb1-71bb77a52b8d","statusText":"","transportId":"4494b3d4-37ba-4d2c-9cb1-71bb77a52b8d"}],"userEq":{},"volume":{"controlType":"attenuation","level":1.0,"muted":false,"stepInterval":0.05000000074505806}},"type":"RECEIVER_STATUS"}
[36mINFO[0m[0000] [application] starting streaming server...
[36mINFO[0m[0000] [application] trying to find available port to start streaming server on
[36mINFO[0m[0000] [application] found available port :49761
[36mINFO[0m[0000] [application] started streaming server
[36mINFO[0m[0000] [application] media server listening on 49761
[36mINFO[0m[0000] [connection] (3)sender-0 -> receiver-0 [urn:x-cast:com.google.cast.receiver]: {"type":"LAUNCH","requestId":3,"appId":"CC1AD845"}
[36mINFO[0m[0000] [connection] (-1)* <- receiver-0 [urn:x-cast:com.google.cast.receiver]: {"requestId":0,"status":{"userEq":{},"volume":{"controlType":"attenuation","level":1.0,"muted":false,"stepInterval":0.05000000074505806}},"type":"RECEIVER_STATUS"}
[36mINFO[0m[0003] [connection] (3)* <- receiver-0 [urn:x-cast:com.google.cast.receiver]: {"requestId":3,"status":{"applications":[{"appId":"CC1AD845","displayName":"Default Media Receiver","isIdleScreen":false,"launchedFromCloud":false,"namespaces":[{"name":"urn:x-cast:com.google.cast.cac"},{"name":"urn:x-cast:com.google.cast.broadcast"},{"name":"urn:x-cast:com.google.cast.media"}],"sessionId":"4022d9ff-0504-4132-b949-2db4dbaedff1","statusText":"Default Media Receiver","transportId":"4022d9ff-0504-4132-b949-2db4dbaedff1"}],"userEq":{},"volume":{"controlType":"attenuation","level":1.0,"muted":false,"stepInterval":0.05000000074505806}},"type":"RECEIVER_STATUS"}
[36mINFO[0m[0003] [connection] (4)sender-0 -> receiver-0 [urn:x-cast:com.google.cast.receiver]: {"type":"GET_STATUS","requestId":4}
[36mINFO[0m[0003] [connection] (4)sender-0 <- receiver-0 [urn:x-cast:com.google.cast.receiver]: {"requestId":4,"status":{"applications":[{"appId":"CC1AD845","displayName":"Default Media Receiver","isIdleScreen":false,"launchedFromCloud":false,"namespaces":[{"name":"urn:x-cast:com.google.cast.cac"},{"name":"urn:x-cast:com.google.cast.broadcast"},{"name":"urn:x-cast:com.google.cast.media"}],"sessionId":"4022d9ff-0504-4132-b949-2db4dbaedff1","statusText":"Default Media Receiver","transportId":"4022d9ff-0504-4132-b949-2db4dbaedff1"}],"userEq":{},"volume":{"controlType":"attenuation","level":1.0,"muted":false,"stepInterval":0.05000000074505806}},"type":"RECEIVER_STATUS"}
[36mINFO[0m[0003] [connection] (5)sender-0 -> 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.tp.connection]: {"type":"CONNECT","requestId":5}
[36mINFO[0m[0003] [connection] (6)sender-0 -> 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"GET_STATUS","requestId":6}
[36mINFO[0m[0003] [connection] (-1)* <- 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"MEDIA_STATUS","status":[],"requestId":0}
[36mINFO[0m[0003] [connection] (6)sender-0 <- 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"MEDIA_STATUS","status":[],"requestId":6}
[36mINFO[0m[0003] [connection] (7)sender-0 -> 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"LOAD","requestId":7,"media":{"contentId":"http://192.168.0.104:49761?media_file=F:\\Media\\test.mkv\u0026live_streaming=false","contentType":"video/mp4","streamType":"BUFFERED","duration":0,"metadata":{"MetadataType":0,"title":"","songName":"","artist":""}},"currentTime":0,"autoplay":true,"queueData":{"startIndex":0},"customData":null}
[36mINFO[0m[0004] [connection] (-1)* <- 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"MEDIA_STATUS","status":[{"mediaSessionId":1,"playbackRate":1,"playerState":"IDLE","currentTime":0,"supportedMediaCommands":274447,"volume":{"level":1,"muted":false},"media":{"contentId":"http://192.168.0.104:49761?media_file=F:\\Media\\test.mkv&live_streaming=false","contentType":"video/mp4","streamType":"BUFFERED","duration":0,"metadata":{"MetadataType":0,"title":"","songName":"","artist":""}},"currentItemId":1,"extendedStatus":{"playerState":"LOADING","media":{"contentId":"http://192.168.0.104:49761?media_file=F:\\Media\\test.mkv&live_streaming=false","contentType":"video/mp4","streamType":"BUFFERED","duration":0,"metadata":{"MetadataType":0,"title":"","songName":"","artist":""}}},"repeatMode":"REPEAT_OFF"}],"requestId":0}
[36mINFO[0m[0004] [application] canServe=true, liveStreaming=false, filename=F:\Media\test.mkv
[36mINFO[0m[0005] [connection] (-1)* <- 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"MEDIA_STATUS","status":[{"mediaSessionId":1,"playbackRate":1,"playerState":"BUFFERING","currentTime":0,"supportedMediaCommands":274447,"volume":{"level":1,"muted":false},"currentItemId":1,"repeatMode":"REPEAT_OFF"}],"requestId":0}
[36mINFO[0m[0005] [connection] (7)* <- 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"MEDIA_STATUS","status":[{"mediaSessionId":1,"playbackRate":1,"playerState":"BUFFERING","currentTime":0,"supportedMediaCommands":274447,"volume":{"level":1,"muted":false},"activeTrackIds":[],"media":{"contentId":"http://192.168.0.104:49761?media_file=F:\\Media\\test.mkv&live_streaming=false","contentType":"video/mp4","streamType":"BUFFERED","duration":2477.517,"metadata":{"MetadataType":0,"title":"","songName":"","artist":""},"tracks":[{"trackId":1,"type":"VIDEO"}],"breakClips":[],"breaks":[]},"currentItemId":1,"items":[{"itemId":1,"media":{"contentId":"http://192.168.0.104:49761?media_file=F:\\Media\\test.mkv&live_streaming=false","contentType":"video/mp4","streamType":"BUFFERED","duration":2477.517,"metadata":{"MetadataType":0,"title":"","songName":"","artist":""}},"autoplay":true,"orderId":0}],"repeatMode":"REPEAT_OFF"}],"requestId":7}
[36mINFO[0m[0006] [connection] (-1)* <- 4022d9ff-0504-4132-b949-2db4dbaedff1 [urn:x-cast:com.google.cast.media]: {"type":"MEDIA_STATUS","status":[{"mediaSessionId":1,"playbackRate":1,"playerState":"PLAYING","currentTime":0.767,"supportedMediaCommands":274447,"volume":{"level":1,"muted":false},"activeTrackIds":[],"currentItemId":1,"repeatMode":"REPEAT_OFF"}],"requestId":0}
[36mINFO[0m[0009] [connection] (-1)Tr@n$p0rt-0 <- Tr@n$p0rt-0 [urn:x-cast:com.google.cast.tp.heartbeat]: {"type":"PING"}`

When I try to run the FFMPEG-command in application.go manually I receive the following output:
`ffmpeg -i "F:\Media\test.mkv" -vcodec "h264" -f "mp4" -movflags "frag_keyframe+faststart" -strict -experimental`

`ffmpeg version N-92544-gb9aff7a53d Copyright (c) 2000-2018 the FFmpeg developers
  built with gcc 8.2.1 (GCC) 20181017
  configuration: --enable-gpl --enable-version3 --enable-sdl2 --enable-fontconfig --enable-gnutls --enable-iconv --enable-libass --enable-libbluray --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libtheora --enable-libtwolame --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libzimg --enable-lzma --enable-zlib --enable-gmp --enable-libvidstab --enable-libvorbis --enable-libvo-amrwbenc --enable-libmysofa --enable-libspeex --enable-libxvid --enable-libaom --enable-libmfx --enable-amf --enable-ffnvcodec --enable-cuvid --enable-d3d11va --enable-nvenc --enable-nvdec --enable-dxva2 --enable-avisynth
  libavutil      56. 24.101 / 56. 24.101
  libavcodec     58. 40.100 / 58. 40.100
  libavformat    58. 23.100 / 58. 23.100
  libavdevice    58.  6.100 / 58.  6.100
  libavfilter     7. 46.101 /  7. 46.101
  libswscale      5.  4.100 /  5.  4.100
  libswresample   3.  4.100 /  3.  4.100
  libpostproc    55.  4.100 / 55.  4.100
Trailing options were found on the commandline.
Input #0, matroska,webm, from 'F:\Media\test.mkv':
  Metadata:
    encoder         : libx264 + libebml + libmatroska + libfastrar
    creation_time   : 2018-04-06T01:13:08.000000Z
  Duration: 00:41:17.52, start: 0.000000, bitrate: 4862 kb/s
    Stream #0:0: Video: h264 (High), yuv420p(tv, bt709/unknown/unknown, progressive), 1280x718, SAR 1:1 DAR 640:359, 23.98 fps, 23.98 tbr, 1k tbn, 47.95 tbc (default)
    Stream #0:1: Audio: ac3, 48000 Hz, 5.1(side), fltp, 384 kb/s (default)`

The version of FFMPEG (Windows) I'm using is:
`ffmpeg -version`

`ffmpeg version N-92544-gb9aff7a53d Copyright (c) 2000-2018 the FFmpeg developers
built with gcc 8.2.1 (GCC) 20181017
configuration: --enable-gpl --enable-version3 --enable-sdl2 --enable-fontconfig --enable-gnutls --enable-iconv --enable-libass --enable-libbluray --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libtheora --enable-libtwolame --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libzimg --enable-lzma --enable-zlib --enable-gmp --enable-libvidstab --enable-libvorbis --enable-libvo-amrwbenc --enable-libmysofa --enable-libspeex --enable-libxvid --enable-libaom --enable-libmfx --enable-amf --enable-ffnvcodec --enable-cuvid --enable-d3d11va --enable-nvenc --enable-nvdec --enable-dxva2 --enable-avisynth
libavutil      56. 24.101 / 56. 24.101
libavcodec     58. 40.100 / 58. 40.100
libavformat    58. 23.100 / 58. 23.100
libavdevice    58.  6.100 / 58.  6.100
libavfilter     7. 46.101 /  7. 46.101
libswscale      5.  4.100 /  5.  4.100
libswresample   3.  4.100 /  3.  4.100
libpostproc    55.  4.100 / 55.  4.100`

**Comment by vishen** [comment]
When I was testing, I was able to get the the chromecast to play `mkv` files by just setting the content type to `mp4` (with no transcoding). I am wondering if this is incorrect.

Are you able to remove the extension from the filename and try running again: `'go-chromecast load "F:\Media\test" -n "Woonkamer" --debug` ? This should force `go-chromecast` to try and transcode to mp4 video.

**Comment by DDoubleU** [comment]
Brilliant, thank you!! This time he's transcoding! Now I've got video and audio, but still no subtitles. Are they "supported" by go-chromecast?

**Comment by vishen** [comment]
I have pushed a fix for this (and added a new release `v0.0.4`). You should be able to run with `go-chromecast load "F:\Media\test.mkv" -n "Woonkamer" --debug` again with the last release :)
